### DISCUSSION TITLE
Setting up FriendAccessAllowed Correctly

### DISCUSSION THREAD

**Discussion by NWoodsman** [dotnet/wpf#9702] [General]
I'm getting many Visual Studio errors when working with the WPF repo. Many of the errors are simple namespace-related issues.

For example, if I open the file `src\PresentationCore\System\Windows\Media\textformatting\TextLineBreak.cs`:

```
The type or namespace name 'SecurityCriticalDataForSet<>' could not be found (are you missing a using directive or an assembly reference?)	
```

Upon investigation, the `SecurityCriticalDataForSet<T>` is implemented in the `Ms.Internal` namespace, and navigating to `src\Shared\MS\Internal\SecurityCriticalDataForSet.cs` shows that the type definition is `internal struct SecurityCriticalDataForSet<T>`.

Note that within `TextLineBreak.cs` I can see Visual Studio warning me of an unused using statement `using Ms.Internal`. The types within Ms.Internal are not being exposed correctly as they are internal. 

grepping the entire repository shows that `InternalsVisibleTo` text is not present anywhere in the repo (I assume WPF predates the invention of this attribute).Therefore some other mechanism is exposing internal types to external assemblies.

I believe it is the `FriendAccessAllowedAttribute`. There must be something missing in my cloned repository that is not resolving correctly.

Can anyone help explain what I need to do to have the internals exposed to friend assemblies correctly.

**Comment by h3xds1nz** [comment]
`InternalsVisibleTo` are set for the assemblies, in the same file as type forwards - `OtherAssemblyAttrs.cs` for each assembly.

Hopefully #6561 is gonna make it soon to get rid of this ugly attribute.

IntelliSense usually works fine but it takes a configuration swap and re-open the file, or if the file is shared in multiple assemblies, just swap to one and then it catches up.
